src/main/kotlin/com/exactpro/th2/common/utils/message/MessageHolder.kt

```diff
@@ -0,0 +1,195 @@
+            }
+        }
+    }
+}
+
+class TransportMessageHolder(
+    val source: ParsedMessage,
+    val book: String,
+    val sessionGroup: String
+) : MessageHolder {
+    override val id: MessageID by lazy { source.id.toProto(book, sessionGroup) }
+    override val parentEventId: EventID? by lazy { source.eventId?.toProto() }
+    override val messageType: String
+        get() = source.type
+    override val protocol: String
+        get() = source.protocol
+    override val properties: Map<String, String>
+        get() = this.source.metadata
+    override val protoMessage: Message by lazy { source.toProto(book, sessionGroup) }
+
+    @Throws(FieldNotFoundException::class)
+    override fun getSimple(vararg path: String): String? = source.getSimple(*path)
+
+    override fun toString(): String {
+        return "TransportMessageWrapper(transport=$source, book='$book', sessionGroup='$sessionGroup')"
+    }
+
+    companion object {
+        /**
+         * Traverses the internal message and returns value by [path]
+         * @return null when the last element exist but has null value otherwise return [Any] value
+         * @throws [FieldNotFoundException] if message doesn't include full path or message structure doesn't match to path
+         */
+        @Throws(FieldNotFoundException::class)
+        fun ParsedMessage.getField(vararg path: String): Any? = runCatching {
+            require(path.isNotEmpty()) {
+                "Path to field can't be empty"
+            }
+            var currentValue: Any? = body[path.first()]
+
+            path.asSequence().drop(1).forEachIndexed { pathIndex, name ->
+                currentValue = when (currentValue) {
+                    is Map<*, *> -> (currentValue as Map<*, *>)[name]
+                    is List<*> -> {
+                        val index = requireNotNull(name.toIntOrNull()) {
+                            "'$name' path element can't be path as number, value: ${currentValue}, path: ${path.contentToString()}, index: ${pathIndex + 1}"
+                        }
+                        val casted = (currentValue as List<*>)
+                        require(index >= 0 && casted.size > index) {
+                            "'$index' index should be positive or zero and less then '${casted.size}' list size, value: ${currentValue}, path: ${path.contentToString()}, index: ${pathIndex + 1}"
+                        }
+                        casted[index]
+                    }
+
+                    else -> error("Field '$name' can't be got from unknown value: ${currentValue}, path: ${path.contentToString()}, index: ${pathIndex + 1}")
+                }
+            }
+            currentValue
+        }.onFailure {
+            throw FieldNotFoundException("Filed not found by ${path.contentToString()} path in $this message", it)
+        }.getOrThrow()
```

Comment:
This is how the `onFailure` looks like inside
```kotlin
public inline fun <T> Result<T>.onFailure(action: (exception: Throwable) -> Unit): Result<T> {
    contract {
        callsInPlace(action, InvocationKind.AT_MOST_ONCE)
    }
    exceptionOrNull()?.let { action(it) }
    return this
}
```
If you throw an exception inside the execution will never get to  `getOrThrow` method.

So, I would suggest using `getOrElse` instead and throwing the exception in the passed lambda.